cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/security/jwt/JwtAuthenticator.java

```diff
@@ -0,0 +1,172 @@
+import javax.servlet.ServletResponse;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.function.Function;
+
+/**
+ * <p>The {@link JwtAuthenticator} adds SSO capabilites to Cruise Control. The expected token is a Json Web Token (JWT).
+ * This class should be used with {@link JwtLoginService} as the token check is carried out by that one. This class
+ * handles redirects for unauthenticated requests and CORS preflight requests.</p>
+ * <p>The workflow can be described with the following diagram:
+ * <pre>
+ *       Client -----1., 4.----> Cruise Control
+ *        |  ^                        |
+ *        |  |________2.______________|
+ *        |
+ *        |
+ *        |------3.------------> Authentication
+ *                                 provider
+ * </pre>
+ * <ol>
+ * <li>The client makes an initial call to Cruise Control
+ * <li>If the request doesn't have a JWT cookie by the specified cookie name, it will be redirected to the authentication
+ *     service to obtain it. If the request is an OPTIONS request we presume it's a CORS preflight request so it'll skip
+ *     the authentication (if the user is authenticated at this point we'll use the existing credentials).
+ * <li>The client authenticates with the provider and obtains the SSO token.
+ * <li>The client can present the JWT cookie to Cruise Control. Cruise Control will validate the cookie with the
+ *    {@link JwtLoginService} by checking its signature, audience and expiration.
+ * </ol>
+ * </p>
+ */
+public class JwtAuthenticator extends LoginAuthenticator {
+
+  public static final String JWT_TOKEN_REQUEST_ATTRIBUTE = "com.linkedin.kafka.cruisecontrol.JwtTokenAttribute";
+  static final Logger JWT_LOGGER = LoggerFactory.getLogger("kafka.cruisecontrol.jwt.logger");
+
+  private static final String METHOD = "JWT";
+  private static final String BEARER = "Bearer";
+
+  private final String _cookieName;
+  private final Function<HttpServletRequest, String> _authenticationProviderUrlGenerator;
+
+  /**
+   * Creates a new {@link JwtAuthenticator} instance with a custom authentication provider url and a cookie name that
+   * will be populated by the authentication service with the JWT token.
+   * @param authenticationProviderUrl is the HTTP(S) address of the authentication service. It will be used to create
+   *                                  the redirection url. For instance <code>https://www.my-auth-service.com/websso?origin={redirectUrl}</code>
+   *                                  will generate <code>https://www.my-auth-service.com/websso?origin=https://www.cruise-control.cc/state</code>
+   *                                  which should redirect from <code>my-auth-service.com</code> to <code>cruise-control.cc/state</code>
+   *                                  after obtaining the JWT token.
```

Comment:
Is `null` an acceptable value for `authenticationProviderUrl`? -- e.g. `JwtSecurityProvider#authenticator()` may set its value to `null`.